diff --git a/dist/lib/isChakraElement.js b/dist/lib/isChakraElement.js
index f627822d23431b0f8a6086eda2f1c402f332abc0..3eca5084c5bbfb788373b15de3a62df069d2b4f7 100644
--- a/dist/lib/isChakraElement.js
+++ b/dist/lib/isChakraElement.js
@@ -16,7 +16,7 @@ function isChakraElement(node, parserServices) {
         return false;
     }
     const specifier = getModuleSpecifierOfImportSpecifier(symbol);
-    return specifier === "@chakra-ui/react";
+    return specifier === "@chakra-ui/react" || specifier === "@screamingdemon/ui";
 }
 // eslint-disable-next-line @typescript-eslint/ban-types -- This Symbol is imported from "typescript"
 function getModuleSpecifierOfImportSpecifier(symbol) {